### ISSUE TITLE
A new option to allow dotnet tool install should return exit code 0 if tool is already installed

### ISSUE THREAD

**Issue by wli3** [dotnet/sdk#10310]
Continue of https://github.com/dotnet/cli/issues/11259 to make the topic more specific. Even thought "dotnet update" with "--version" could solve the similar issue, but it is hard to discover.

**Comment by Lonli-Lokli** [comment]
What is the timeline for this option? It's really surprising that multiple installations exits with code 1.
You know, even installers do not do that.

**Comment by wli3** [comment]
Sorry I don't have a timeline

**Comment by augustoproiete** [comment]
Would creating a separate command make it easier to get this implemented sooner? E.g. `dotnet ensure <toolname>` that would:

- Do nothing if the tool & exact version is installed, and returns `0` exit code
- Installs the tool if the tool is not yet installed, and returns `0` exit code
- Updates the tool to the specified version, if a different version is already installed, and returns `0` exit code

e.g.

```powershell
dotnet tool ensure --global Cake.Tool --version 0.38.5
```

- If Cake.Tool is not installed, install it specifically with version 0.38.5
- If Cake.Tool is already installed, but a lower version, automatically upgrade to 0.38.5
- If Cake.Tool is already installed, but a higher version, automatically downgrade to 0.38.5

All cases would return `0` as exit code. The only reason for returning a non-zero exit code would be failure to download the tool, write to disk, etc. Exceptions...

---

A common scenario is CI builds where you just want to make sure that the tools you depend on for the build are present.

**Comment by rafaelrruiz** [comment]
Any news on this?  The option has been discussed for almost two years.

**Comment by Lonli-Lokli** [comment]
@rafaelrruiz you should use dotnet tool restore instead, it will install if deps is missed and skip if already installed.
It requires newer SDK

**Comment by augustoproiete** [comment]
> @rafaelrruiz you should use dotnet tool restore instead, it will install if deps is missed and skip if already installed.
> It requires newer SDK

@Lonli-Lokli This is really not the same thing that is being asked. To use `dotnet tool restore`, you'd need to have installed the tool beforehand and have a tool manifest in the specified directory. The goal here is to have an idempotent way to install .NET tools with a single command that Just Work™.

**Comment by Lonli-Lokli** [comment]
@augustoproiete it cannot be idempotent even by your definition - first installation should not only install dep, but create tools manifest, while second run not.

There is no silver bullet for now, there was a discussion before and honestly speaking it's better to install dependency first with dotnet install locally, save manifest in got and use dotnet restore after.

**Comment by augustoproiete** [comment]
@Lonli-Lokli I'm not sure I follow. Creating the manifest automatically (if installing as local tool) is one of the possible missing steps that would make it idempotent - which is the end goal. And, of course, if installing as a global tool, no manifest needed - but still fails today if tool already exists so another missing piece.

`dotnet restore` is a workaround for some scenarios only. .NET tools is lacking the ability to have a simple command that does exactly what it's being asked: Ensure that a .NET tool is installed and exit gracefully.

The current state leave us with poor a UX:

![image](https://user-images.githubusercontent.com/177608/118564379-8b718b00-b746-11eb-95e6-4b06eb5541cf.png)
https://www.nuget.org/packages/dotnet-reportgenerator-globaltool/

You should be able to copy/paste the instructions on how to install a .NET tool, and it should work on any machine regardless if the tool is installed or not, manifest or not... But that's not where we are today.

**Comment by hotchkj** [comment]
It's deeply disappointing that all of these related issues have been present, discussed, and unaddressed for two or more years. The scenarios outlined appear to be legitimate use-cases for tooling. The Cake example is a perfect one. As I come to finally try and get off the ancient and now-unsupported Cake bootstrapper, I immediately run into the problems outlined in this and the linked issues (and frankly have no idea which one to comment on - they all describe the same pain - so apologies if this belongs elsewhere).

> it's better to install dependency first with dotnet install locally, save manifest in git and use dotnet restore after

I believe this misses the point - nobody wants to update countless Git repos with tooling versions, that's a maintenance nightmare - CI is typically centrally managed, has to be local, and due to the unique way tool manifests are designed, can't just be copied in at CI build time either, as usual build tools would be.

I just want my NET tools installed locally for CI. I don't want to update & maintain 200+ repos with manifest files that can get out of sync in yet another tooling ecosystem. I don't want to pollute build agents with global state, and global tools also just don't work for some cases. I just want the tool installed idempotently. Why is this challenging? The NET SDK itself recognised this immediately, and gave us `dotnet-install.ps1`. Instant local NET SDK, job done. Where is a local NET tool equivalent?

If I have an existing Git repo with zero or more tool manifests
- I can't just run `dotnet new tool-manifest`, it'd either fail or overwrite the existing ones if they're there, and the exit code doesn't tell me anything
- Only the NET SDK seems to know what is and isn't a valid tool manifest, so I can't even write reliable code to detect it
  - Also why make the decision that it has to be a hardcoded path & name, yet give us a `--tool-manifest` option in `install` that will happily take any filename?!?
  - This is not documented, either
- I need to install a CI-only local tool - but `install` will fail if there's no manifest or if the tool is already present, with unhelpful exit codes
- `update` uninstalls and reinstalls the tool, instead of just ensuring the current version is present, requiring re-download time

I would ask the question more generally as to why the wheel has been reinvented here (again). With NuGet & PackageReference, for example, there's a nice standard convention of 'request package, package gets cached in location that is either global or specified by environment variable, package is installed where requested if it isn't already present'. `dotnet tool` does not satisfy any of these idempotency criteria today.

**Comment by nabeelio** [comment]
This behavior also breaks Azure Pipelines if you're using Task Groups several times in one job (with different parameters) which has a tool install step

**Comment by ChristopherHaws** [comment]
We had to tell our Azure Pipeline to ignore errors in this task because it returns an error code =(

```yaml
- task: DotNetCoreCLI@2
  displayName: install sqlpackage
  # dotnet tool install returns error code 1 if the package is already installed =(
  # https://github.com/dotnet/sdk/issues/10310
  continueOnError: true
  inputs:
    command: custom
    custom: tool
    arguments: install -g microsoft.sqlpackage
```

I would be fine with there being an option to ignore this error to prevent it from being a breaking change, but it really should just be the default behavior. The flag _should_ be if you want it to throw if already installed...

**Comment by JL03-Yue** [comment]
#35088 return exit code 0 if tool is installed when using `dotnet tool install`

**Comment by baronfel** [comment]
I'm going to close this one because the linked PR merged in 8.0.2xx. Thanks for the feedback!
